### PR TITLE
CDAP-4286: HTTP handler for ACL management

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-4286
http://builds.cask.co/browse/CDAP-DUT3236

### CODE REVIEW

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/CheckAuthorizedRequest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * Request for checking if a user can perform certain actions on an entity.
+ */
+public class CheckAuthorizedRequest {
+
+  private final EntityId entity;
+  private final String user;
+  private final Set<Action> actions;
+
+  public CheckAuthorizedRequest(EntityId entity, String user, Set<Action> actions) {
+    Preconditions.checkArgument(entity != null, "enttiy is required");
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/CheckAuthorizedRequest.java]: misspelling. `entity`

File: cdap-proto/src/test/java/co/cask/cdap/proto/security/CheckAuthorizedRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ */
+public class CheckAuthorizedRequestTest {
+
+  @Test
+  public void testValidation() {
+    new CheckAuthorizedRequest(Ids.namespace("foo"), "bob", ImmutableSet.of(Action.READ));
+
+    try {
+      new CheckAuthorizedRequest(Ids.namespace("foo"), null, null);
```
Comment (anwar6953) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/CheckAuthorizedRequestTest.java]: put an
`Assert.fail` before the `catch`, if the exception is expected to happen?

Comment (anwar6953) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/CheckAuthorizedRequestTest.java]: same for all the expected assertions in this test case.

Comment (anwar6953) [inline cdap-proto/src/test/java/co/cask/cdap/proto/security/CheckAuthorizedRequestTest.java]: Same in `RevokeRequestTest` and `GrantRequestTest`.

File: cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ */
+public class RevokeRequest {
+
+  private final EntityId entity;
+  private final String user;
+  private final Set<Action> actions;
+
+  public RevokeRequest(EntityId entity, @Nullable String user, @Nullable Set<Action> actions) {
+    Preconditions.checkArgument(entity != null, "entity is required");
+    Preconditions.checkArgument(actions == null || (user != null), "actions is required when user is ");
```
Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: Are you requiring that if `user` is non null, then `actions` also has to be non-null?
The condition `actions == null | user != null` doesn't seem to ensure that.
For instance, `user` could be `non-null`, `actions` could be `null`, and the condition could still return true.

Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: Nope, I want to only allow the following:

entity: non-null, user: non-null, actions: non-null
entity: non-null, user: non-null, actions: null
entity: non-null, user: null, actions: null

Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: It looks like the 2nd case you listed is not consistent with the error message.

Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: Oops, it should say `actions is required when user is provided`. Will fix.

Comment (jwang47) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: Wait, nevermind. Should say `user is required when action is provided`..

Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: The 2nd case still seems inconsistent with that message.
User is provided, but actions is `null`.

Comment (anwar6953) [inline cdap-proto/src/main/java/co/cask/cdap/proto/security/RevokeRequest.java]: Ok, that last message seems appropriate:
`user is required when action is provided`

**Comment by gokulavasan** [comment]
@jwang47 Can we add the HTTP endpoints' description in the JIRA?

**Comment by jwang47** [comment]
@gokulavasan Added.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java [hunk 4]
```diff
@@ -142,7 +147,7 @@ protected int getInstances(HttpRequest request) throws IllegalArgumentException,
     Reader reader = new InputStreamReader(new ChannelBufferInputStream(content), Charsets.UTF_8);
     try {
       return GSON.fromJson(reader, type);
-    } catch (JsonSyntaxException e) {
+    } catch (Exception e) {
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: Any reason for this change?

Comment (jwang47) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/util/AbstractAppFabricHttpHandler.java]: `GSON.fromJson()` can fail with a `RuntimeException` due to a missing no-args constructor, and it's useful to see the stack trace from the test run. Can change to `RuntimeException`.

File: cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  public boolean authorized(EntityId entity, String user, Set<Action> actions)
+    throws IOException, UnauthorizedException {
+
+    CheckAuthorizedRequest checkRequest = new CheckAuthorizedRequest(entity, user, actions);
+
+    URL url = config.resolveURLV3("security/authorized");
+    HttpRequest request = HttpRequest.post(url).withBody(GSON.toJson(checkRequest)).build();
+    HttpResponse response = restClient.execute(request, config.getAccessToken());
+
+    CheckAuthorizedResponse checkResponse = GSON.fromJson(
```
Comment (gokulavasan) [inline cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java]: Looks like the response is a simple boolean? Wondering, if we need a new class.

Comment (jwang47) [inline cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java]: This is using a class -- the response looks like `{"authorized": true}`.

Comment (gokulavasan) [inline cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java]: Ah ok, right. Ignore my comment.

**Comment by gokulavasan** [comment]
Reviewed. Address @anwar6953 's comments and should be LGTM.

**Comment by gokulavasan** [comment]
Question, if the user doesn't choose the default auth plugin, we shouldn't add the AuthHandler right?

**Comment by jwang47** [comment]
AuthorizationHandler should only be added when authorization is enabled, since it only relies on AuthorizationPlugin. Any implementation of AuthorizationPlugin will work according to the contract.

**Comment by gokulavasan** [comment]
Changes looks good. @anwar6953 any comments?

**Comment by anwar6953** [comment]
lgtm

**Comment by jwang47** [comment]
Tested on cluster. Will merge once build passes.

**Pull request merged**